src/recommendations/src/recommendations-service/app.py

```diff
@@ -286,122 +320,217 @@ def recommendations():
+            user_id=user_id,
+            item_list=unranked_items,
+            tracker=tracker
+        )
+
+        app.logger.debug(f"Experiment ranking resolver gave us this ranking: {ranked_items}")
+
+        resp_headers['X-Experiment-Name'] = experiment.name
+        resp_headers['X-Experiment-Type'] = experiment.type
+        resp_headers['X-Experiment-Id'] = experiment.id
+    else:
+        # Fallback to default behavior of checking for campaign ARN parameter and
+        # then the default product resolver.
+        values = get_parameter_values([default_campaign_arn_param_name, filter_purchased_param_name])
+        app.logger.info(f'Falling back to Personalize: {values}')
+
+        campaign_arn = values[0]
+        filter_arn = values[1]
+
+        if campaign_arn:
+            resolver = PersonalizeRankingResolver(campaign_arn=campaign_arn, filter_arn=filter_arn)
+            resp_headers['X-Personalize-Recipe'] = get_recipe(campaign_arn)
+        else:
+            app.logger.info(f'Falling back to No-op: {values}')
+            resolver = RankingProductsNoOpResolver()
+
+        ranked_items = resolver.get_items(
+            user_id=user_id,
+            product_list=unranked_items
+        )
+
+    response_items = []
+    for ranked_item in ranked_items:
+
+        # Unlike with /recommendations and /related we are not hitting the products API to get product info back
+        # One would hope the caller would have it since the products are coming from there
+        item = item_map.get(ranked_item.get('itemId'))
+
+        if 'experiment' in ranked_item:
+
+            # We do not want to override an experiment if info already there
+            # e.g. if client is pulling recommendations then reranking them for example
+            # - this should not happen as the first recommendation pull would already be ranked
+            if 'experiment' not in item:
+                # The client may just use the correlation ID but here
+                # we give full information about the experiment to the client anyway
+                item['experiment'] = ranked_item['experiment']
+
+            if 'url' in item:
+                # Append the experiment correlation ID to the product URL so it gets tracked if used by client.
+                product_url = item.get('url')
+                if '?' in product_url:
+                    product_url += '&'
                 else:
-                    resolver = RankingProductsNoOpResolver()
+                    product_url += '?'
 
-                ranked_items = resolver.get_items(
-                    user_id = user_id, 
-                    product_list = unranked_items
-                )
+                product_url += 'exp=' + ranked_item['experiment']['correlationId']
 
-            response_items = []
-            for ranked_item in ranked_items:
-                item = item_map.get(ranked_item.get('itemId'))
+                item['old_url'] = item['url']
```

Comment:
In the existing reranking code, which we endeavoured to keep as intact as possible considering we also repurposed it for discounts, the item URL is retrieved and the experiment ID added. This is presumably so that downstream it can be seen and picked up (though client-side mostly the experiment dictionary is used). We attempted to remove references to the experiment if the product was not involved in the discount recommendation but clearly the better approach overall would be to leave it in. Another approach would be to introduce into the experimentation a separate resolver type to handle with the "picking" type experiment.